pkg/server/image_pull.go

```diff
@@ -184,19 +216,37 @@ func (c *criContainerdService) pullImage(ctx context.Context, ref string) (
 	}
 	// TODO(random-liu): What if following operations fail? Do we need to do cleanup?
 
+	resources := newResourceSet()
+
+	glog.V(4).Infof("Start downloading resources for image %q", ref)
 	// Fetch all image resources into content store.
 	// Dispatch a handler for a sequence of handlers which:
-	// 1) fetch the object using a FetchHandler;
-	// 2) recurse through any sub-layers via a ChildrenHandler.
+	// 1) track all resources associated using a customized handler;
+	// 2) fetch the object using a FetchHandler;
+	// 3) recurse through any sub-layers via a ChildrenHandler.
 	err = containerdimages.Dispatch(
 		ctx,
 		containerdimages.Handlers(
+			containerdimages.HandlerFunc(func(ctx gocontext.Context, desc imagespec.Descriptor) (
+				[]imagespec.Descriptor, error) {
+				resources.add(remotes.MakeRefKey(ctx, desc))
+				return nil, nil
+			}),
 			remotes.FetchHandler(c.contentStoreService, fetcher),
 			containerdimages.ChildrenHandler(c.contentStoreService)),
 		desc)
 	if err != nil {
-		return "", "", fmt.Errorf("failed to fetch image %q desc %+v: %v", ref, desc, err)
+		// Dispatch returns error when requested resources are locked.
+		// In that case, we should start waiting and checking the pulling
```

Comment:
> @mikebrow so they are not always sync calls... :-) This explains some of the errors I was seeing... that the concurrent request may not be finished yet thus the locked resource!

It is sync call. The problem is that it returns error if someone else holds a lock of a resource.